flamedisx/nest/lxe_blocks/quanta_splitting.py

```diff
@@ -2,110 +2,476 @@
 o = tf.newaxis
 
 
 @export
-class MakePhotonsElectronsBinomial(fd.Block):
+class MakePhotonsElectronsNR(fd.Block):
+    is_ER = False
 
-    do_pel_fluct = False
-
-    depends_on = ((('quanta_produced',), 'rate_vs_quanta'),)
     dimensions = ('electrons_produced', 'photons_produced')
+    bonus_dimensions = (('ions_produced', True),)
+    depends_on = ((('energy',), 'rate_vs_energy'),)
+
+    max_dim_size = {'ions_produced': 30}
+
+    exclude_data_tensor = ('ions_produced_min', 'ions_produced_max')
 
-    special_model_functions = ('p_electron',)
+    special_model_functions = ('mean_yields', 'recomb_prob', 'skewness', 'variance',
+                               'width_correction', 'mu_correction')
     model_functions = special_model_functions
 
-    p_electron = 0.5   # Nonsense, ER and NR sources provide specifics
+    use_batch = True
 
     def _compute(self,
                  data_tensor, ptensor,
+                 #
+                 i_batch,
                  # Domain
                  electrons_produced, photons_produced,
+                 # Bonus dimension
+                 ions_produced,
                  # Dependency domain and value
-                 quanta_produced, rate_vs_quanta):
-        pel = self.source.gimme('p_electron', bonus_arg=quanta_produced,
-                                data_tensor=data_tensor, ptensor=ptensor)
-
-        # Create tensors with the dimensions of our final result
-        # i.e. (n_events, |photons_produced|, |electrons_produced|),
-        # containing:
-        # ... numbers of total quanta produced
+                 energy, rate_vs_energy):
+
+        def compute_single_energy_full(args):
+            # Compute the block for a single energy, without approximations
+
+            energy = args[0]
+            rate_vs_energy = args[1]
+            ions_min = args[2]
+
+            ions_min = tf.repeat(ions_min[:, o], tf.shape(ions_produced)[1], axis=1)
+            ions_min = tf.repeat(ions_min[:, :, o], tf.shape(ions_produced)[2], axis=2)
+            ions_min = tf.repeat(ions_min[:, :, :, o], tf.shape(ions_produced)[3], axis=3)
+
+            # Calculate the ion domain tensor for this energy
+            _ions_produced = ions_produced_add + ions_min
+
+            if self.is_ER:
+                nel_mean = self.gimme('mean_yield_electron', data_tensor=data_tensor, ptensor=ptensor,
+                                      bonus_arg=energy)
+                nq_mean = self.gimme('mean_yield_quanta', data_tensor=data_tensor, ptensor=ptensor,
+                                     bonus_arg=(energy, nel_mean))
+                fano = self.gimme('fano_factor', data_tensor=data_tensor, ptensor=ptensor,
+                                  bonus_arg=nq_mean)
+
+                p_nq = tfp.distributions.Normal(loc=nq_mean,
+                                                scale=tf.sqrt(nq_mean * fano) + 1e-10).cdf(nq + 0.5) \
+                    - tfp.distributions.Normal(loc=nq_mean,
```

Comment:
Same distribution constructed twice here. The 'else' has the same issue twice

Reply:
Done.